File: backend/internal/models/ai.go
package models

// ProviderType identifies the API dialect a provider speaks.
type ProviderType string

const (
	ProviderOpenAICompatible ProviderType = "openai-compatible" // OpenAI, DeepSeek, Ollama, etc.
	ProviderAnthropic        ProviderType = "anthropic"
	ProviderGemini           ProviderType = "gemini" // reserved; not yet implemented
)

// AIProvider is a configured LLM provider (OpenAI, DeepSeek, Ollama, ...).
type AIProvider struct {
	ID          string       `json:"id"`
	Name        string       `json:"name"`
	DisplayName string       `json:"display_name"`
	BaseURL     string       `json:"base_url"`
	APIKeyEnc   string       `json:"-"`            // encrypted at rest, never serialized
	APIKeyHint  string       `json:"api_key_hint"` // e.g. "sk-...abcd" for the UI
	APIType     ProviderType `json:"api_type"`
	IsEnabled   bool         `json:"is_enabled"`
	IsDefault   bool         `json:"is_default"`
	CreatedAt   string       `json:"created_at"`
	UpdatedAt   string       `json:"updated_at"`
}

// AIModel is a model endpoint exposed by a provider.
type AIModel struct {
	ID             string `json:"id"`
	ProviderID     string `json:"provider_id"`
	Name           string `json:"name"`
	ModelID        string `json:"model_id"`
	ModelType      string `json:"model_type"` // chat | embedding | image
	MaxTokens      int    `json:"max_tokens"`
	SupportsStream bool   `json:"supports_stream"`
	SupportsTools  bool   `json:"supports_tools"`
	IsEnabled      bool   `json:"is_enabled"`
	IsDefault      bool   `json:"is_default"`
	CreatedAt      string `json:"created_at"`
	UpdatedAt      string `json:"updated_at"`
}

File: frontend/src/services/ai.ts
import { apiGet, apiPost, apiPut, apiDelete } from './api';

export interface AIProvider {
    id: string;
    name: string;
    display_name: string;
    base_url: string;
    api_key_hint?: string;
    api_type: 'openai-compatible' | 'anthropic' | 'gemini' | 'mock';
    is_enabled: boolean;
    is_default: boolean;
    created_at?: string;
    updated_at?: string;
}

export interface AIModel {
    id: string;
    provider_id: string;
    name: string;
    model_id: string;
    model_type: string;
    max_tokens: number;
    supports_stream: boolean;
    supports_tools: boolean;
    is_enabled: boolean;
    is_default: boolean;
    created_at?: string;
    updated_at?: string;
}

export interface ChatMessage {
    role: 'system' | 'user' | 'assistant';
    content: string;
}

export const aiApi = {
    listProviders: () => apiGet<{ code: number; data: AIProvider[] }>('/ai/providers').then(r => r.data),

    getProvider: (id: string) =>
        apiGet<{ code: number; data: AIProvider }>(`/ai/providers/${id}`).then(r => r.data),

    createProvider: (provider: Partial<AIProvider> & { api_key_enc?: string }) =>
        apiPost<{ code: number; data: AIProvider }>('/ai/providers', provider).then(r => r.data),

    updateProvider: (id: string, provider: Partial<AIProvider> & { api_key_enc?: string }) =>
        apiPut<{ code: number; data: AIProvider }>(`/ai/providers/${id}`, provider).then(r => r.data),

    deleteProvider: (id: string) =>
        apiDelete<{ code: number }>(`/ai/providers/${id}`),

    listModels: (providerId: string) =>
        apiGet<{ code: number; data: AIModel[] }>(`/ai/models/${providerId}`).then(r => r.data),

    chat: async function* (providerId: string, modelId: string, messages: ChatMessage[]): AsyncGenerator<string> {
        const token = localStorage.getItem('token') || 'dev-token-1234';
        const resp = await fetch(`/api/v1/ai/chat`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${token}`,
            },
            body: JSON.stringify({
                provider_id: providerId,
                model_id: modelId,
                messages,
            }),
        });

        if (!resp.ok || !resp.body) {
            throw new Error(`Chat request failed: ${resp.status}`);
        }

        const reader = resp.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            buffer += decoder.decode(value, { stream: true });

            const lines = buffer.split('\n');
            buffer = lines.pop() || '';

            for (const line of lines) {
                const trimmed = line.trim();
                if (trimmed.startsWith('data: ')) {
                    const data = trimmed.slice(6).trim();
                    if (data === '[DONE]') return;
                    if (data.startsWith('{')) {
                        try {
                            const parsed = JSON.parse(data);
                            if (parsed.content) yield parsed.content;
                        } catch {
                            // ignore malformed
                        }
                    }
                }
            }
        }
    },
};